modbus: support async modbus serial

lanwer50 opened the pull request:
Support the use of async api in modbus_serial.c. Reason for this extension are crc mismatches happening due to rx data loss when using the interrupt driven api (default modbus implementation) at higher speeds and/or high cpu load. It is recommended to use the async api (implemented in this MR) along with hardware byte counting requiring a timer instance and one PPI channel to achieve a robust communication at high baudrates. A new Kconfig CONFIG_MODBUS_SERIAL_ASYNC_API to enable/disable this feature has been added. 

Fixes #71767

Signed-off-by: Anouar Raddaoui <a.raddaoui@smight.com>

bjarki-trackunit commented on subsys/modbus/Kconfig:
```diff
@@ -47,6 +47,13 @@ config MODBUS_SERIAL
 	  Enable Modbus over serial line support.
+	  
+config MODBUS_SERIAL_ASYNC_API
+	depends on MODBUS_SERIAL
+    bool "Modbus over serial line with async api support"
+    default n
```
It also depends on `UART_ASYNC_API`, and `default n` is the default so it can be left out :)

lanwer50 replied:
Thanks for the feedback. Changes are done.

MaureenHelm commented:
@jfischer-no please take a look

kartben commented:
@jfischer-no can you please help review when you have time? Thanks!

jfischer-no commented on subsys/modbus/modbus_serial.c:
```diff
@@ -592,11 +657,20 @@ int modbus_serial_init(struct modbus_context *ctx,
+			LOG_ERR("uart_callback_set returns error = %d", err);
+		}
+	} else {
+		uart_irq_callback_user_data_set(cfg->dev, uart_cb_handler, ctx);
```
Please check return value here as well.
```c
		err = uart_irq_callback_user_data_set(cfg->dev, uart_cb_handler, ctx);
		if (!err) {
			k_timer_init(&cfg->rtu_timer, rtu_tmr_handler, NULL);
			k_timer_user_data_set(&cfg->rtu_timer, ctx);
			modbus_serial_rx_on(ctx);
		}
```

jfischer-no commented on subsys/modbus/modbus_serial.c:
```diff
@@ -592,11 +657,20 @@ int modbus_serial_init(struct modbus_context *ctx,
-	modbus_serial_rx_on(ctx);
 	LOG_INF("RTU timeout %u us", cfg->rtu_timeout);
 
 	return 0;
```
`return err;`

lanwer50 replied:
Done

jfischer-no reviewed (changes requested):
The ASYNC part is not tested, the default testing platform does not support ASYNC mode. I need to find some time to find out how to resolve it.

lanwer50 commented:
I would like to remove the "closed" label to adapt the changes suggested in the review.

lanwer50 commented on subsys/modbus/modbus_serial.c:
```diff
@@ -37,14 +37,27 @@ static void modbus_serial_tx_on(struct modbus_context *ctx)
+		/* Do nothing */
+	} else {
+		uart_irq_tx_disable(cfg->dev);
+	}
```
Done

lanwer50 commented on subsys/modbus/modbus_serial.c:
```diff
@@ -58,14 +71,31 @@ static void modbus_serial_rx_on(struct modbus_context *ctx)
+
+		if (err) {
+			LOG_ERR("uart_rx_enable returns error = %d", err);
+		}
```
Done

lanwer50 commented on subsys/modbus/modbus_serial.c:
```diff
@@ -401,6 +431,41 @@ static void uart_cb_handler(const struct device *dev, void *app_data)
+	case UART_RX_DISABLED:
+		/* Do nothing */
+		break;
+	}
```
Done

lanwer50 commented on subsys/modbus/modbus_serial.c:
```diff
@@ -592,11 +657,20 @@ int modbus_serial_init(struct modbus_context *ctx,
-	k_timer_init(&cfg->rtu_timer, rtu_tmr_handler, NULL);
-	k_timer_user_data_set(&cfg->rtu_timer, ctx);
+	if (IS_ENABLED(CONFIG_MODBUS_SERIAL_ASYNC_API)) {
+		int err = uart_callback_set(cfg->dev, uart_cb_async_handler, ctx);
```
Done

lanwer50 commented on subsys/modbus/modbus_serial.c:
```diff
@@ -592,11 +657,20 @@ int modbus_serial_init(struct modbus_context *ctx,
+
+		if (err != 0) {
+			LOG_ERR("uart_callback_set returns error = %d", err);
+		}
```
Done

jfischer-no commented on subsys/modbus/Kconfig:
```diff
@@ -48,6 +48,12 @@ config MODBUS_SERIAL
+	depends on MODBUS_SERIAL && UART_ASYNC_API
+	bool "Modbus over serial line with async api support"
+	help
+	  Enable Modbus Serial to use Async API.
```
Maybe "Use UART ASYNC API for Modbus over serial line"?

lanwer50 replied:
Changed

jfischer-no commented on subsys/modbus/modbus_serial.c:
```diff
@@ -70,14 +82,30 @@ static void modbus_serial_rx_on(struct modbus_context *ctx)
+	if (IS_ENABLED(CONFIG_MODBUS_SERIAL_ASYNC_API)) {
+		int err;
+
+		memset(cfg->uart_buf, 0, sizeof(cfg->uart_buf));
```
Why is there memset()?

lanwer50 replied:
I think you're right, there is no need for it. I just tend to clean up buffers used before by tx. 
However, this should not be needed -> removed.

lanwer50 replied:
I think you're right, there is no need for it. I just tend to clean up buffers used before by tx. 
However, this should not be needed -> removed.

jfischer-no commented on subsys/modbus/modbus_serial.c:
```diff
@@ -424,6 +452,40 @@ static void uart_cb_handler(const struct device *dev, void *app_data)
+		cfg->uart_buf_ctr = evt->data.rx.len;
+		k_work_submit(&ctx->server_work);
+		break;
+	case UART_TX_ABORTED:
+	case UART_RX_STOPPED:
+	case UART_RX_BUF_REQUEST:
+	case UART_RX_BUF_RELEASED:
+	case UART_RX_DISABLED:
```
```c
	case UART_TX_ABORTED:
		__fallthrough;
	case UART_RX_STOPPED:
		__fallthrough;
	case UART_RX_BUF_REQUEST:
		__fallthrough;
	case UART_RX_BUF_RELEASED:
		__fallthrough;
	case UART_RX_DISABLED:
```

Commit:
modbus: support UART ASYNC API in modbus serial

Reason for this extension are CRC mismatches happening due to RX data
loss when using the interrupt driven API at higher speeds and/or high
CPU load. It is recommended to use the ASYNC API along with hardware
byte counting requiring a timer instance and one PPI channel to achieve
a robust communication at high baudrates on nRF platforms.

Signed-off-by: Anouar Raddaoui <a.raddaoui@smight.com>

lanwer50 commented on subsys/modbus/Kconfig:
Done

lanwer50 commented:
> The ASYNC part is not tested, the default testing platform does not support ASYNC mode. I need to find some time to find out how to resolve it.

@jfischer-no: I think the easiest way to test it out will be to use two nrf52840 dk_boards with the default modbus samples, one as client and the second board as server. Then, add the new Kconfig MODBUS_SERIAL_ASYNC_API to the client prf.conf sample.

jfischer-no reviewed (approved):
Thanks
